bitcoin/src/blockdata/block.rs

```diff
@@ -90,150 +84,203 @@ impl Decodable for Version {
         }
-
-        false
     }
 
-    /// Computes the transaction Merkle root.
-    pub fn compute_merkle_root(&self) -> Option<TxMerkleNode> {
-        let hashes = self.txdata.iter().map(|obj| obj.compute_txid());
-        TxMerkleNode::calculate_root(hashes)
-    }
+    (false, None)
+}
 
-    /// Computes the witness commitment for the block's transaction list.
-    pub fn compute_witness_commitment(
-        witness_root: WitnessMerkleNode,
-        witness_reserved_value: &[u8],
-    ) -> WitnessCommitment {
-        let mut encoder = sha256d::Hash::engine();
-        witness_root.consensus_encode(&mut encoder).expect("engines don't error");
-        encoder.input(witness_reserved_value);
-        WitnessCommitment::from_byte_array(sha256d::Hash::from_engine(encoder).to_byte_array())
-    }
+fn witness_commitment_from_coinbase(coinbase: &Transaction) -> Option<WitnessCommitment> {
+    // Consists of OP_RETURN, OP_PUSHBYTES_36, and four "witness header" bytes.
+    const MAGIC: [u8; 6] = [0x6a, 0x24, 0xaa, 0x21, 0xa9, 0xed];
 
-    /// Computes the Merkle root of transactions hashed for witness.
-    pub fn witness_root(&self) -> Option<WitnessMerkleNode> {
-        let hashes = self.txdata.iter().enumerate().map(|(i, t)| {
-            if i == 0 {
-                // Replace the first hash with zeroes.
-                Wtxid::COINBASE
-            } else {
-                t.compute_wtxid()
-            }
-        });
-        WitnessMerkleNode::calculate_root(hashes)
+    if !coinbase.is_coinbase() {
+        return None;
     }
 
-    /// Returns the weight of the block.
-    ///
-    /// > Block weight is defined as Base size * 3 + Total size.
-    pub fn weight(&self) -> Weight {
-        // This is the exact definition of a weight unit, as defined by BIP-141 (quote above).
-        let wu = self.base_size() * 3 + self.total_size();
-        Weight::from_wu_usize(wu)
+    // Commitment is in the last output that starts with magic bytes.
+    if let Some(pos) = coinbase
+        .output
+        .iter()
+        .rposition(|o| o.script_pubkey.len() >= 38 && o.script_pubkey.as_bytes()[0..6] == MAGIC)
+    {
+        let bytes =
+            <[u8; 32]>::try_from(&coinbase.output[pos].script_pubkey.as_bytes()[6..38]).unwrap();
+        Some(WitnessCommitment::from_byte_array(bytes))
+    } else {
+        None
     }
+}
 
-    /// Returns the base block size.
+/// Extension functionality for the [`Block<Checked>`] type.
+pub trait BlockCheckedExt: sealed::Sealed {
+    /// Constructs a new [`Block`].
     ///
-    /// > Base size is the block size in bytes with the original transaction serialization without
-    /// > any witness-related data, as seen by a non-upgraded node.
-    fn base_size(&self) -> usize {
-        let mut size = Header::SIZE;
+    /// # Returns
+    ///
+    /// Return the block if it is valid, `None` if not. See [`Block::validate`].
+    #[allow(clippy::new_ret_no_self)] // This does return `Self`.
+    fn new(
```

Comment:
In 186e3c797e410661db36564aeab4ae3972bb8a70:

We should call this `new_checked` for symmetry with `new_unchecked`.